tests/simulator_testfixture.h

```diff
@@ -346,123 +345,159 @@ void TestApplyGate5() {
 
 template <typename Simulator>
 void TestApplyControlGate() {
-    constexpr char circuit_str[] =
-R"(6
-0 h 0
-0 h 1
-0 h 2
-0 h 3
-0 h 4
-0 h 5
-1 c 0 t 1
-2 rx 0 0.1
-2 ry 1 0.2
-2 rz 2 0.3
-2 rx 3 0.4
-2 ry 4 0.5
-2 rz 5 0.6
-3 c 0 1 h 2
-4 rz 0 0.7
-4 rx 1 0.8
-4 ry 2 0.9
-4 rz 3 1.0
-4 rx 4 1.1
-4 rx 5 1.2
-5 c 0 1 2 t 3
-6 ry 0 1.3
-6 rz 1 1.4
-6 rx 2 1.5
-6 ry 3 1.6
-6 rz 4 1.7
-6 rx 5 1.8
-7 c 0 1 2 3 t 4
-8 rz 0 1.9
-8 rx 1 2.0
-8 ry 2 2.1
-8 rz 3 2.2
-8 rx 4 2.3
-8 ry 5 2.4
-9 c 0 is 1 2
-10 rx 0 2.5
-10 ry 1 2.6
-10 rz 2 2.7
-10 rx 3 2.8
-10 ry 4 2.9
-10 rz 5 3.0
-11 c 0 1 is 2 3
-12 ry 0 3.1
-12 rz 1 3.2
-12 rx 2 3.3
-12 ry 3 3.4
-12 rz 4 3.5
-12 rx 5 3.6
-13 c 0 1 2 is 3 4
-14 rz 0 3.7
-15 rx 1 3.8
-15 ry 2 3.9
-15 rz 3 4.0
-15 rx 4 4.1
-15 rx 5 4.2
-16 c 0 1 2 3 is 4 5
-17 rx 0 4.3
-17 ry 1 4.4
-17 rz 2 4.5
-17 rx 3 4.6
-17 ry 4 4.7
-17 rz 5 4.8
-18 c 3 is 4 5
-19 ry 0 4.9
-19 rz 1 5.0
-19 rx 2 5.1
-19 ry 3 5.2
-19 rz 4 5.3
-19 rx 5 5.4
-20 c 4 is 0 1
-21 rz 0 5.5
-21 rx 1 5.6
-21 ry 2 5.7
-21 rz 3 5.8
-21 rx 4 5.9
-21 ry 5 6.0
-22 c 4 is 0 2
-23 rx 0 6.1
-23 ry 1 6.2
-23 rz 2 6.3
-23 rx 3 6.4
-23 ry 4 6.5
-23 rz 5 6.6
-24 c 4 is 0 5
-25 ry 0 6.7
-25 rz 1 6.8
-25 rx 2 6.9
-25 ry 3 7.0
-25 rz 4 7.1
-25 rx 5 7.2
-26 c 4 h 5
-27 rz 0 7.3
-27 rx 1 7.4
-27 ry 2 7.5
-27 rz 3 7.6
-27 rx 4 7.7
-27 ry 5 7.8
-)";
-
   using StateSpace = typename Simulator::StateSpace;
   using fp_type = typename StateSpace::fp_type;
-
-  Circuit<GateQSim<fp_type>> circuit;
-  std::stringstream ss(circuit_str);
-  EXPECT_TRUE(CircuitQsimParser<IO>::FromStream(99, "string", ss, circuit));
-  EXPECT_EQ(circuit.num_qubits, 6);
-  EXPECT_EQ(circuit.gates.size(), 97);
+  using Gate = GateQSim<fp_type>;
+
+  unsigned num_qubits = 6;
+
+  std::vector<Gate> gates;
+  gates.reserve(128);
+
+  gates.push_back(GateHd<fp_type>::Create(0, 0));
+  gates.push_back(GateHd<fp_type>::Create(0, 1));
+  gates.push_back(GateHd<fp_type>::Create(0, 2));
+  gates.push_back(GateHd<fp_type>::Create(0, 3));
+  gates.push_back(GateHd<fp_type>::Create(0, 4));
+  gates.push_back(GateHd<fp_type>::Create(0, 5));
+
+  gates.push_back(GateT<fp_type>::Create(1, 1));
+  MakeControlledGate({0}, {1}, gates.back());
```

Comment:
If `MakeControlledGate` returned the generated gate, this sequence of `push_back`s could be replaced with one big vector-initialization, which would parallel nicely with Cirq's Circuit constructors.

Reply:
This is a good point.